### PR TITLE
feat(gcp): add provider id validation inside test_connection

### PR DESCRIPTION
**Pull request by pedrooot**
### Description

This PR adds a validation inside test_connection in order to check that the `provider_id` provided is reachable for the given credentials.

### Checklist

- Are there new checks included in this PR? Yes / No
    - If so, do we need to update permissions for the provider? Please review this carefully.
- [x] Review if the code is being covered by tests.
- [x] Review if code is being documented following this specification https://github.com/google/styleguide/blob/gh-pages/pyguide.md#38-comments-and-docstrings
- [x] Review if backport is needed.

### License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: prowler/providers/gcp/exceptions/exceptions.py [hunk 1]
```diff
@@ -37,6 +37,10 @@ class GCPBaseException(ProwlerException):
             "message": "Error loading static credentials",
             "remediation": "Check the credentials and ensure they are properly set up. client_id, client_secret and refresh_token are required.",
         },
+        (1933, "GCPNotValidProviderIdError"): {
```
Comment (jfagoagas) [inline prowler/providers/gcp/exceptions/exceptions.py]:
```suggestion
        (1933, "GCPInvalidAccountCredentials"): {
```

The point here is to validate that the provided credentials does not match with the `project_id` that wants to be used.

File: prowler/providers/gcp/gcp_provider.py [hunk 2]
```diff
@@ -528,3 +550,28 @@ def validate_static_arguments(
...
             "type": "authorized_user",
         }
+
+    @staticmethod
+    def validate_project_id(provider_id: str, credentials: str = None) -> bool:
+        """
+        Validate the provider ID
+
+        Args:
+            provider_id: str
```
Comment (jfagoagas) [inline prowler/providers/gcp/gcp_provider.py:559]: Please improve the docstring with:
- Better explanation of what the function does
- Args and their description

File: prowler/providers/gcp/gcp_provider.py [hunk 3]
```diff
@@ -316,9 +324,16 @@ def test_connection(
...
                 )
 
-            session, _ = GcpProvider.setup_session(
+            session, project_id = GcpProvider.setup_session(
                 credentials_file, service_account, gcp_credentials
             )
+
+            if provider_id and project_id != provider_id:
+                # Logic to check if the provider ID matches the project ID
+                GcpProvider.validate_project_id(
```
Comment (jfagoagas) [inline prowler/providers/gcp/gcp_provider.py:332]: I think it'll be easier if you include the code within the function, so it does not match raise Exception and if matches return `connnected = True`. What do you think?

File: prowler/providers/gcp/gcp_provider.py (hunk 2 above)
Comment (jfagoagas) [reply prowler/providers/gcp/gcp_provider.py:559]: Nothing to do if you remove it.

File: prowler/providers/gcp/gcp_provider.py [hunk 4]
```diff
@@ -316,9 +324,16 @@ def test_connection(
                     client_id, client_secret, refresh_token
                 )
 
-            session, _ = GcpProvider.setup_session(
+            session, project_id = GcpProvider.setup_session(
                 credentials_file, service_account, gcp_credentials
             )
+
+            if provider_id and project_id != provider_id:
```
Comment (jfagoagas) [inline prowler/providers/gcp/gcp_provider.py:330]:
```suggestion
            if provider_id and project_id != provider_id and provider_id not in GcpProvider.get_projects(credentials=credentials):
```

but my question is why do we need to get all the projects if the setup session returns the project id.

**Review by jfagoagas — CHANGES_REQUESTED** [review]
Thanks @pedrooot, I left some questions/changes. How can we add tests for this code?

File: prowler/providers/gcp/exceptions/exceptions.py (hunk 1 above)
Comment (pedrooot) [reply prowler/providers/gcp/exceptions/exceptions.py]: Done!

File: prowler/providers/gcp/gcp_provider.py (hunk 2 above)
Comment (pedrooot) [reply prowler/providers/gcp/gcp_provider.py:559]: DOne!

File: prowler/providers/gcp/gcp_provider.py (hunk 4 above)
Comment (pedrooot) [reply prowler/providers/gcp/gcp_provider.py:330]: The setup_session returns the default project_id, could happen that the given one doesn't match with the default

File: prowler/providers/gcp/gcp_provider.py (hunk 3 above)
Comment (pedrooot) [reply prowler/providers/gcp/gcp_provider.py:332]: I would prefer to modularize the code into the `validate_project_id` method, this way I can document what the method does in order to clarify the usage for the user.

**Review by jfagoagas — APPROVED** [review]
Great one 👏

**Pull request merged**